crates/ursa-gateway/src/worker/cache.rs

```diff
@@ -0,0 +1,106 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use async_trait::async_trait;
+use tokio::sync::{mpsc::UnboundedSender, oneshot};
+use tracing::log::warn;
+
+use crate::cache::Tlrfu;
+
+pub struct Cache {
+    tlrfu: Tlrfu,
+    tx: UnboundedSender<WorkerCacheCommand>,
+}
+
+impl Cache {
+    pub fn new(max_size: u64, ttl_buf: u128, tx: UnboundedSender<WorkerCacheCommand>) -> Self {
+        Self {
+            tlrfu: Tlrfu::new(max_size, ttl_buf),
+            tx,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum WorkerCacheCommand {
+    GetSync {
+        key: String,
+    },
+    InsertSync {
+        key: String,
+        value: Arc<Vec<u8>>,
+    },
+    Fetch {
+        cid: String,
+        sender: oneshot::Sender<Result<Arc<Vec<u8>>>>,
+    },
+    TtlCleanUp,
+}
+
+#[async_trait]
+pub trait WorkerCache: Send + Sync + 'static {
+    async fn get(&mut self, k: &str) -> Result<()>;
+    async fn insert(&mut self, k: String, v: Arc<Vec<u8>>) -> Result<()>;
+    async fn ttl_cleanup(&mut self) -> Result<()>;
+}
+
+#[async_trait]
+impl WorkerCache for Cache {
+    async fn get(&mut self, k: &str) -> Result<()> {
+        self.tlrfu.get(&String::from(k)).await?;
+        Ok(())
+    }
+
+    async fn insert(&mut self, k: String, v: Arc<Vec<u8>>) -> Result<()> {
+        if !self.tlrfu.contains(&k) {
+            self.tlrfu.insert(k, v).await?;
+        } else {
+            warn!("[Cache]: attempt to insert existed key: {k}");
+        }
+        Ok(())
+    }
+
+    async fn ttl_cleanup(&mut self) -> Result<()> {
+        self.tlrfu.process_tll_clean_up().await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+pub trait ServerCache: Send + Sync + 'static {
+    async fn get_announce(&self, k: &str) -> Result<Arc<Vec<u8>>>;
+}
+
+#[async_trait]
+impl ServerCache for Cache {
+    async fn get_announce(&self, k: &str) -> Result<Arc<Vec<u8>>> {
+        if let Some(data) = self.tlrfu.dirty_get(&String::from(k)) {
+            self.tx.send(WorkerCacheCommand::GetSync {
+                key: String::from(k),
+            })?;
+            Ok(Arc::clone(data))
+        } else {
+            let (tx, rx) = oneshot::channel();
+            self.tx.send(WorkerCacheCommand::Fetch {
+                cid: String::from(k),
+                sender: tx,
+            })?;
+            let data = rx.await??;
```

Comment:
@scott-dn I don't think this model of [send request to node] -> [read response and cache it] -> [send respond] is not going to work with stream piping the IO (client <-> node). Do you have some ideas?

Reply:
Yeah, with cache, we don't have to deal with stream data at the moment (we might want to stream from cache).
But for the case of files larger than 1GB, we don't want to cache them; we should stream all the way instead.

Reply:
How would we know the size before requesting content from the cache nodes or are you thinking that we will use `content-lenght`?

Reply:
block info will do.
> https://github.com/fleek-network/ursa/pull/158#discussion_r1057999388

Reply:
@theBeardA so what does the "block info" rpc method return?

Reply:
we dont have "block info" method yet, what I meant in the comment was blockchain info and not info about data.
but if you want to know the size of the data, the root block have that information I think. so maybe we can use ursa_get_cid to get the root block and get the info from that @kckeiks

Reply:
@theBeardA hmm, interesting; maybe cache node RPC can provide `block info`?
If that's too much, gateway can inspect the root block for the information.

Reply:
I don't think it's going to be too much to add this rpc method on the cache node.
but isn't this something resolver should be doing? just a design choice.
lmk what you think @scott-dn @b0xtch @kckeiks

Reply:
If we have to make a call to inspect a root block just to get size info about the content, that's another RTT.

Reply:
We could inspect HTTP headers for things such as `Range` or`transfer-encoding: chunk` to indicate intent of download a large file then pipe the io.

Reply:
> If we have to make a call to inspect a root block just to get size info about the content, that's another RTT.

It will be fine if we can have a cache to track content size history.


> We could inspect HTTP headers, from the client request, for things such as `Range` or`transfer-encoding: chunk` to indicate intent to download a large file then pipe the io.

Sounds good.

Reply:
How about we include the file size in the metadata?

Reply:
> How about we include the file size in the metadata?

isn't it another round trip to the indexer?

Reply:
I don't think so. Aren't we currently calling the global indexer, given a cid? That returns the metadata. So instead of adding another rpc method or calling ursa_get_cid to get the file size, we could add the file size in the metadata. Thoughts?

Reply:
> > isn't it another round trip to the indexer?
> 
> I don't think so. Aren't we currently calling the global indexer, given a cid? That returns the metadata. So instead of adding another rpc method or calling ursa_get_cid to get the file size, we could add the file size in the metadata. Thoughts?

sounds good to me.

Reply:
That would be a lot easier then messing around with HTTP headers 😅

Reply:
Agree 💯 we should have an issue ticket for this. 🙏

Reply:
Here https://github.com/fleek-network/ursa/issues/206

Left some questions, notes, links from research.

Reply:
https://github.com/fleek-network/ursa/issues/207